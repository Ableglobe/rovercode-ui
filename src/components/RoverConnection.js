import React, { Component, Fragment } from 'react';
import { Card, Label, Icon } from 'semantic-ui-react';
import { hot } from 'react-hot-loader';
import PropTypes from 'prop-types';
import Websocket from 'react-websocket';

const heartbeatTimeout = 8000; // milliseconds

class RoverConnection extends Component {
  constructor(props) {
    super(props);

    this.timeout = null;
    this.state = {
      online: false,
    };
  }

  componentDidMount() {
    this.startHeartbeatTimer();
  }

  startHeartbeatTimer = () => {
    this.timeout = setTimeout(this.setOffline, heartbeatTimeout);
  }

  setActive = () => {
    const { changeActiveRover, clientId } = this.props;

    changeActiveRover(clientId);
  }

  setOffline = () => this.setState({ online: false })

  setOnline = () => this.setState({ online: true })

  onMessage = (data) => {
    const { changeLeftSensorState, changeRightSensorState } = this.props;
    const { online } = this.state;

    const message = JSON.parse(data);

    if (message.type === 'sensor-reading') {
      const value = message['sensor-value'];
      if (message['sensor-id'] === 'ultrasonic-left') {
        changeLeftSensorState(value);
      } else if (message['sensor-id'] === 'ultrasonic-right') {
        changeRightSensorState(value);
      }
    } else if (message.type === 'heartbeat') {
      clearTimeout(this.timeout);
      this.startHeartbeatTimer();

      if (!online) {
        this.setOnline();
      }
    }
  }

  render() {
    const { clientId, isActive, name } = this.props;
    const { online } = this.state;

    const wsProtocol = window.location.protocol === 'https:' ? 'wss' : 'ws';
    const wsUrl = `${wsProtocol}://${window.location.hostname}/ws/realtime/${clientId}`;

    return (
      <Fragment>
        <Card color={isActive ? 'blue' : null} onClick={this.setActive}>
          <Card.Content>
            <Label corner="right" style={{ borderColor: 'white' }}>
              {
                online ? (
                  <Icon name="circle" color="green" />
                ) : (
                  <Icon name="circle outline" />
                )
              }
            </Label>
            <Card.Header>
              { name }
            </Card.Header>
            <Card.Meta>
              { isActive ? 'Active' : 'Inactive' }
            </Card.Meta>
          </Card.Content>
        </Card>
        <Websocket url={wsUrl} onMessage={this.onMessage} />
      </Fragment>
    );
  }
}

RoverConnection.propTypes = {
  isActive: PropTypes.bool.isRequired,
  clientId: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  changeLeftSensorState: PropTypes.func.isRequired,
  changeRightSensorState: PropTypes.func.isRequired,
  changeActiveRover: PropTypes.func.isRequired,
};

export default hot(module)(RoverConnection);
