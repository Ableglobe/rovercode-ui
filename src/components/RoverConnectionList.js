import React, { Component, Fragment } from 'react';
import { Card, Loader } from 'semantic-ui-react';
import { hot } from 'react-hot-loader';
import PropTypes from 'prop-types';

import RoverConnection from '@/components/RoverConnection';

class RoverConnectionList extends Component {
  componentDidMount() {
    const { fetchRovers } = this.props;

    return fetchRovers();
  }

  render() {
    const {
      activeRover,
      changeActiveRover,
      changeLeftSensorState,
      changeRightSensorState,
      commands,
      popCommand,
      rovers,
    } = this.props;

    let activeRoverObject = null;
    let inactiveRovers = null;

    if (rovers) {
      const activeRoverIndex = rovers.findIndex(rover => rover.client_id === activeRover);

      if (activeRoverIndex > -1) {
        ({ [`${activeRoverIndex}`]: activeRoverObject, ...inactiveRovers } = rovers);
        inactiveRovers = Object.values(inactiveRovers);
      }
    }

    if (!inactiveRovers) {
      inactiveRovers = rovers;
    }

    return (
      <Fragment>
        {
          rovers ? (
            <Card.Group style={{ margin: '10px' }}>
              {
                activeRoverObject ? (
                  <RoverConnection
                    key={activeRoverObject.id}
                    clientId={activeRoverObject.client_id}
                    name={activeRoverObject.name}
                    changeLeftSensorState={changeLeftSensorState}
                    changeRightSensorState={changeRightSensorState}
                    changeActiveRover={changeActiveRover}
                    popCommand={popCommand}
                    commands={commands}
                    isActive
                  />
                ) : (null)
              }
              {
                inactiveRovers.map(rover => (
                  <RoverConnection
                    key={rover.id}
                    clientId={rover.client_id}
                    name={rover.name}
                    changeLeftSensorState={changeLeftSensorState}
                    changeRightSensorState={changeRightSensorState}
                    changeActiveRover={changeActiveRover}
                    popCommand={popCommand}
                    isActive={false}
                  />
                ))
              }
            </Card.Group>
          ) : (
            <Loader active />
          )
        }
      </Fragment>
    );
  }
}

RoverConnectionList.defaultProps = {
  activeRover: null,
  rovers: null,
};

RoverConnectionList.propTypes = {
  activeRover: PropTypes.string,
  rovers: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      client_id: PropTypes.string.isRequired,
    }),
  ),
  fetchRovers: PropTypes.func.isRequired,
  changeLeftSensorState: PropTypes.func.isRequired,
  changeRightSensorState: PropTypes.func.isRequired,
  changeActiveRover: PropTypes.func.isRequired,
  popCommand: PropTypes.func.isRequired,
  commands: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default hot(module)(RoverConnectionList);
