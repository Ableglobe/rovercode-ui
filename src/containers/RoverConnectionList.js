import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { hot } from 'react-hot-loader';
import { withCookies, Cookies } from 'react-cookie';

import {
  changeActiveRover as actionChangeActiveRover,
  popCommand as actionPopCommand,
  fetchRovers,
} from '@/actions/rover';
import {
  changeLeftSensorState as actionChangeLeftSensorState,
  changeRightSensorState as actionChangeRightSensorState,
} from '@/actions/sensor';
import { updateValidAuth } from '@/actions/auth';
import RoverConnectionList from '@/components/RoverConnectionList';

const mapStateToProps = ({ rover }) => ({ ...rover });
const mapDispatchToProps = (dispatch, { cookies }) => ({
  fetchRovers: () => {
    const fetchRoversAction = fetchRovers({
      headers: {
        Authorization: `JWT ${cookies.get('auth_jwt')}`,
      },
    });
    return dispatch(fetchRoversAction).catch((error) => {
      if (error.response.status === 401) {
        // Authentication is no longer valid
        dispatch(updateValidAuth(false));
      }
    });
  },
  changeLeftSensorState: state => dispatch(actionChangeLeftSensorState(state)),
  changeRightSensorState: state => dispatch(actionChangeRightSensorState(state)),
  changeActiveRover: clientId => dispatch(actionChangeActiveRover(clientId)),
  popCommand: () => dispatch(actionPopCommand()),
});

const RoverConnectionListContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(RoverConnectionList);

RoverConnectionListContainer.propTypes = {
  cookies: PropTypes.instanceOf(Cookies).isRequired,
};

export default hot(module)(withCookies(RoverConnectionListContainer));
