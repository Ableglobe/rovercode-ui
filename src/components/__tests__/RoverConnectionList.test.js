import React from 'react';
import { MemoryRouter } from 'react-router';
import { Card, Loader } from 'semantic-ui-react';
import { shallow, mount } from 'enzyme';
import RoverConnection from '../RoverConnection';
import RoverConnectionList from '../RoverConnectionList';

let changeActiveRover;
let changeLeftSensorState;
let changeRightSensorState;
let commands;
let popCommand;
let fetchRovers;

describe('The RoverConnectionList component', () => {
  beforeEach(() => {
    changeActiveRover = jest.fn();
    changeLeftSensorState = jest.fn();
    changeRightSensorState = jest.fn();
    commands = [];
    popCommand = jest.fn();
    fetchRovers = jest.fn(() => Promise.resolve({}));
  });

  test('renders on the page with no errors', () => {
    const wrapper = shallow(
      <RoverConnectionList
        changeActiveRover={changeActiveRover}
        changeLeftSensorState={changeLeftSensorState}
        changeRightSensorState={changeRightSensorState}
        popCommand={popCommand}
        fetchRovers={fetchRovers}
        commands={commands}
        rovers={null}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('fetches rovers on mount', async () => {
    await mount(
      <MemoryRouter>
        <RoverConnectionList
          changeActiveRover={changeActiveRover}
          changeLeftSensorState={changeLeftSensorState}
          changeRightSensorState={changeRightSensorState}
          popCommand={popCommand}
          fetchRovers={fetchRovers}
          commands={commands}
        />
      </MemoryRouter>,
    );
    expect(fetchRovers.mock.calls.length).toBe(1);
  });

  test('shows loading during rover fetch', () => {
    const wrapper = shallow(
      <RoverConnectionList
        changeActiveRover={changeActiveRover}
        changeLeftSensorState={changeLeftSensorState}
        changeRightSensorState={changeRightSensorState}
        popCommand={popCommand}
        fetchRovers={fetchRovers}
        commands={commands}
        isFetching
      />,
    );
    expect(wrapper.find(Card).exists()).toBe(false);
    expect(wrapper.find(Loader).exists()).toBe(true);
  });

  test('shows the correct rovers for the user', async () => {
    const rovers = {
      next: null,
      previous: null,
      results: [{
        id: 1,
        name: 'Sparky',
        client_id: '1234',
      }, {
        id: 2,
        name: 'Marvin',
        client_id: '5678',
      }],
    };
    const wrapper = shallow(
      <RoverConnectionList
        activeRover="5678"
        changeActiveRover={changeActiveRover}
        changeLeftSensorState={changeLeftSensorState}
        changeRightSensorState={changeRightSensorState}
        popCommand={popCommand}
        fetchRovers={fetchRovers}
        rovers={rovers}
        commands={commands}
      />,
    );
    await wrapper.instance().componentDidMount();
    wrapper.update();

    expect(wrapper.find(Loader).exists()).toBe(false);
    expect(wrapper.find(RoverConnection).length).toBe(2);
    expect(wrapper.find(RoverConnection).first().prop('name')).toBe('Marvin');
    expect(wrapper.find(RoverConnection).last().prop('name')).toBe('Sparky');
  });
});
