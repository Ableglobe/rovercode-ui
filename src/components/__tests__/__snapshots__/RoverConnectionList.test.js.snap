// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The RoverConnectionList component renders on the page with no errors 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RoverConnectionList
    activeRover={null}
    changeActiveRover={[MockFunction]}
    changeLeftSensorState={[MockFunction]}
    changeRightSensorState={[MockFunction]}
    commands={Array []}
    fetchRovers={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": Promise {},
          },
        ],
      }
    }
    isFetching={false}
    popCommand={[MockFunction]}
    rovers={null}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <Loader
        active={true}
      />,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "active": true,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <Loader
          active={true}
        />,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "active": true,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
